Hw02 express

Paulina-Maciesowicz opened the pull request.

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
u

u

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
update

update

Paulina-Maciesowicz committed:
update

update readme

RadziuMM commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,78 @@
+router.get("/:id", async (req, res, next) => {
+  const id = req.params.id;
+  const contactsId = await contactsMethots.getContactById(id);
+  console.log(id);
```
Powyrzucaj wszystkie console.logi które nic nie wnoszą, czyli zostaw praktycznie tylko wyświetnie błędów aplikacji

RadziuMM commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,78 @@
-  res.json({ message: 'template message' })
-})
+router.post("/", async (req, res, next) => {
+  const { name, email, phone } = req.body;
```
Zróbmy tutaj walidację, pominiesz wtedy wyżucanie i łapanie błędów z funkcji + wtedy też sprawdzi ci czy są odpowiednie pola oraz czy nie ma dodatkowych

```
router.post("/", async (req, res, next) => {
	try {
		const validation = schema.validate(req.body);
		if (validation.error) {
			res.status(400).json({ error: validation.error.message });
		}
		const contactsAdded = await contactsMethots.addContact(req.body);
		res.status(201).json(contactsAdded);
	} catch (error) {
		console.error("Error adding contact:", error);
		res.status(500).json({ error: "Internal server error" });
	}
});
```

i wtedy addContacts powinno wyglądać tak

```
async function addContact({ name, email, phone }) {
	const data = await fs.readFile(contactsPath, "utf-8");
	const contacts = JSON.parse(data);
	const lastContact = contacts[contacts.length - 1];
	const newId = parseInt(lastContact.id) + 1;
	const newContact = { id: newId.toString(), name, email, phone };
	contacts.push(newContact);
	await fs.writeFile(contactsPath, JSON.stringify(contacts));
	return newContact;
}
```
Tylko wtedy musisz przenieść scheme tutaj, albo dać ją do innego pliku i zaimportować. Zrób to samo dla reszty endpointów bo raczej to rzucanie błędów jest średnim pomysłem

RadziuMM commented on routes/api/contactsMethots.js:
```diff
@@ -0,0 +1,108 @@
+    return newContact;
+  } catch (error) {
+    console.error(error);
+    return `Validation error: ${validation.error.message}`;
```
Tu już pisałem że do zmiany, ale tu tak dla informacji powiem że obsługa tego try catcha nie zadziała jak powinien. Jeśli validacja będzie pomyślna a wystąpi błąd np w zapisie pliku to zwrócisz 201 z błędem jako string (i chyba nawet wyskoczy ci błąd przy res.status(201).json("Validation error.......")  bo to string a nie json)

Paulina-Maciesowicz committed:
update

1. Przeniesiona walidacja
2. Usuniętę console.log'i

RadziuMM commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,82 @@
-router.get('/', async (req, res, next) => {
-  res.json({ message: 'template message' })
-})
+router.get("/", async (req, res, next) => {
```
Chyba średnio to wytłumaczyłem ostatnio przy tym postcie (post już jest ok, ale chciał bym abyś przerzuciła te try/catche) z contact methods tutaj. Możliwe że ten await jest skomplikowany ale on przeciąga błędy do funkcji nadrzędnej.

```
 router.get("/", async (req, res, next) => {
  try {
      const contacts = await contactsMethots.listContacts();
      res.status(200).json(contacts);
  } catch (error) {
     console.error(error);
     res.status(500).json({  message: "Unknown error occurred" });
  } 
});

```

```
async function listContacts() {
    const data = await fs.readFile(contactsPath, "utf-8");
    return JSON.parse(data);
}
```

RadziuMM commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,82 @@
-  res.json({ message: 'template message' })
-})
+router.get("/:id", async (req, res, next) => {
+  const id = req.params.id;
```
```
  try {
    const id = req.params.id;
    const contact = await contactsMethots.getContactById(id);
    if (!contact) {
      res.status(404).json({ message: "Contact not found" });
    }
    
    res.status(200).json(contact);
  } catch (error) {
     console.error(error);
     res.status(500).json({  message: "Unknown error occurred" });
  } 
```

```
async function getContactById(contactId) {
    const data = await fs.readFile(contactsPath, "utf-8");
    const contacts = JSON.parse(data);
    const contact = contacts.find((c) => c.id === contactId);
    return contact;
}
```

RadziuMM commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,82 @@
-router.delete('/:contactId', async (req, res, next) => {
-  res.json({ message: 'template message' })
-})
+router.delete("/:contactId", async (req, res, next) => {
```
```
  const contactId = req.params.contactId;
  try {
    const contact = await contactsMethots.getContactById(contactId);
    if (!contact) {
      return res.status(404).json({ error: "Contact not found" });
    }
    await contactsMethots.removeContact(contactId);
    res.status(200).json({ message: "Contact deleted" });
  } catch (error) {
    console.error("Error deleting contact:", error);
    res.status(500).json({ error: "Internal server error" });
  }
  ```
  ```
  async function removeContact(contactId) {
    const data = await fs.readFile(contactsPath, "utf-8");
    const contacts = JSON.parse(data);
    const updatedContacts = contacts.filter((contact) => contact.id !== contactId);
    const updatedContactsJSON = JSON.stringify(updatedContacts);

    await fs.writeFile(contactsPath, updatedContactsJSON);
}
  ```

RadziuMM commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,82 @@
-router.put('/:contactId', async (req, res, next) => {
-  res.json({ message: 'template message' })
-})
+router.put("/:contactId", async (req, res, next) => {
```
```
  const contactId = req.params.contactId;
  const body = req.body;
  try {
    const validation = schema.validate(req.body);
    if (validation.error) {
      res.status(400).json({ error: validation.error.message });
    }
    const contact = await contactsMethots.getContactById(contactId);
    if (!contact) {
      return res.status(404).json({ error: "Contact not found" });
    }

    await contactsMethots.updateContact(contactId, body);
    res.status(200).json({ message: "Contact updated" });
  } catch (error) {
    console.error("Error deleting contact:", error);
    res.status(500).json({ error: "Internal server error" });
  }
  ```
  
  ```
  async function updateContact(contactId, body) {
    const data = await fs.readFile(contactsPath, "utf-8");
    const contacts = JSON.parse(data);
    const contact = contacts.findIndex((c) => c.id === contactId);
    contacts[contact] = { ...contacts[contact], ...body };
    const updatedContactsJSON = JSON.stringify(contacts);

    await fs.writeFile(contactsPath, updatedContactsJSON);
}
```

RadziuMM reviewed (changes requested):
Ok, prawdopodbnie słabo przy tym postcie to wytłumaczyłem, poprawiłem te try catche i funkcje, zmień sobie i zobacz jak to działa. Raz masz dużo mniej kodu 2 jest dużo prostszy. Gdy coś było nie jasne możesz do mnie napisać na slacku (ogarnianie błędów i await są skomplikowone) chętnie wytłumaczę.

Paulina-Maciesowicz committed:
Update try/catch to contacts.js

Update try/catch to contacts.js

1. Dodałam do 40 linijki -> (router.post("/", async (req, res, next) => ) również return; (45 linijka) ponieważ źle działała walidacja. Mimo wyrzucanych błędów o braku @ lub imienia i tak po późniejszym sprawdzeniu całej listy osób, były te osoby dodawane. Chciałam dopytać dlaczego tak się dzieje ;)?

RadziuMM commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,91 @@
+    const id = req.params.id;
+    const contact = await contactsMethots.getContactById(id);
+    if (!contact) {
+      res.status(404).json({ message: "Contact not found" });
```
o tu też ten return trzeba dać, to o czym pisałem na lms

RadziuMM reviewed (approved):
Jeden return do dodania, zaliczam już, co do pytania odpisałem na lms (jak było by nadal to niejasne to możesz pytać na slacku).

Paulina-Maciesowicz committed:
update return

update return
